drivers/kscan/kscan_npcx.c

```diff
@@ -0,0 +1,514 @@
+	 * changes.
+	 */
+	for (int c = 0; c < CONFIG_KSCAN_NPCX_COLUMN_SIZE; c++) {
+		/* Check if there was an update from the previous scan */
+		row_changed = matrix_new_state[c] ^ data->matrix_previous_state[c];
+
+		if (!row_changed) {
+			continue;
+		}
+
+		for (int r = 0; r < CONFIG_KSCAN_NPCX_ROW_SIZE; r++) {
+			/*
+			 * Index all they keys that changed for each row
+			 * in order to debounce each key in terms of it
+			 */
+			if (row_changed & BIT(r)) {
+				data->scan_cycle_idx[c][r] = data->scan_cycles_idx;
+			}
+		}
+
+		data->matrix_unstable_state[c] |= row_changed;
+		data->matrix_previous_state[c] = matrix_new_state[c];
+	}
+
+	for (int c = 0; c < CONFIG_KSCAN_NPCX_COLUMN_SIZE; c++) {
+		deb_col = data->matrix_unstable_state[c];
+
+		if (!deb_col) {
+			continue;
+		}
+
+		/* Debouncing for each row key occurs here */
+		for (int r = 0; r < CONFIG_KSCAN_NPCX_ROW_SIZE; r++) {
+			uint8_t mask = BIT(r);
+			uint8_t row_bit = matrix_new_state[c] & mask;
+
+			/* Continue if we already debounce a key */
+			if (!(deb_col & mask)) {
+				continue;
+			}
+
+			/* Convert the clock cycle differences to usec */
+			uint32_t debt = k_cyc_to_us_floor32(
+				cycles_now - data->scan_clk_cycle[data->scan_cycle_idx[c][r]]);
+
+			/* Does the key requires more time to be debounced? */
+			if (debt < (row_bit ? data->deb_time_press : data->deb_time_rel)) {
+				/* Need more time to debounce */
+				continue;
+			}
+
+			data->matrix_unstable_state[c] &= ~row_bit;
+
+			/* Check if there was a change in the stable state */
+			if ((data->matrix_stable_state[c] & mask) == row_bit) {
+				/* Key state did not change */
+				continue;
+			}
+
+			/*
+			 * The current row has been debounced, therefore update
+			 * the stable state. Then, proceed to notify the
+			 * application about the keys pressed.
+			 */
+			data->matrix_stable_state[c] ^= mask;
+			if (data->callback) {
+				data->callback(dev, r, c, row_bit ? true : false);
+			}
+		}
+	}
+
+	return key_pressed;
+}
+
+static void kscan_matrix_polling_thread(const struct device *dev, void *dummy2, void *dummy3)
+{
+	struct kscan_npcx_data *const data = dev->data;
+	uint32_t current_cycles;
+	uint32_t cycles_diff;
+	uint32_t wait_period;
+
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	while (true) {
+		/* Enable interrupt of KSI pins */
+		kscan_npcx_resume_detection(dev, true);
+
+		kscan_npcx_drive_column(dev, KEYBOARD_COLUMN_DRIVE_ALL);
+		k_sem_take(&data->poll_lock, K_FOREVER);
+		LOG_DBG("Start KB scan!!");
+
+		/* Disable interrupt of KSI pins and start polling */
+		kscan_npcx_resume_detection(dev, false);
+
+		uint64_t poll_time_end = sys_clock_timeout_end_calc(K_USEC(data->poll_timeout));
+
+		while (atomic_get(&data->enable_scan) == 1U) {
+			uint32_t start_period_cycles = k_cycle_get_32();
+
+			if (check_key_events(dev)) {
+				poll_time_end =
+					sys_clock_timeout_end_calc(K_USEC(data->poll_timeout));
+
+			} else if (k_uptime_ticks() > poll_time_end) {
+				break;
```

Comment:
Can't quite follow this logic but, this is skippin the sleep right? Is there a condition where this spins without yielding?

Reply:
Opps, there is something wrong here. It should check `start_period_cycles > poll_time_end` which likes what is done in [CROS-EC](https://source.chromium.org/chromiumos/chromiumos/codesearch/+/main:src/platform/ec/common/keyboard_scan.c;l=997).
This is used to check if it can leave from poll mode to interrupt mode after no key press for the poll_timeout_us period.